File: python3/098.ValidateBinarySearchTree/ValidateBinarySearchTree.py
# O(n)
# Runtime Beats 100.00%

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        self.prev = None
        return self.inOrder(root)

    def inOrder(self,root:TreeNode) -> bool:
        if root is None:
            return True
        if not self.inOrder(root.left):
            return False
        if self.prev and self.prev.val >= root.val:
            return False
        self.prev = root
        return self.inOrder(root.right)

File: go/098.ValidateBinarySearchTree/ValidateBinarySearchTree.go
// O(n)
// Runtime Beats 100.00%

//Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var pre *TreeNode = nil

func isValidBST(root *TreeNode) bool {
	pre = nil
	return inOrder(root)
}

func inOrder(root *TreeNode) bool {
	if root == nil {
		return true
	}

	if !inOrder(root.Left) {
		return false
	}

	if pre != nil && pre.Val >= root.Val {
		return false
	}
	pre = root
	return inOrder(root.Right)
}
